riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java

```diff
@@ -0,0 +1,183 @@
+import static org.zalando.riptide.Navigators.series;
+import static org.zalando.riptide.Route.responseEntityOf;
+import static org.zalando.riptide.RoutingTree.dispatch;
+import static org.zalando.riptide.stream.Streams.streamOf;
+
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Stream;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpStatus.Series;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.http.client.ClientHttpResponse;
+import org.springframework.stereotype.Component;
+import org.zalando.problem.ThrowableProblem;
+import org.zalando.riptide.Binding;
+import org.zalando.riptide.Bindings;
+import org.zalando.riptide.Rest;
+import org.zalando.riptide.ThrowingConsumer;
+import org.zalando.riptide.capture.Capture;
+
+import com.google.common.io.ByteStreams;
+
+@Component
+public class NakadiGateway {
+
+    public static final MediaType PROBLEM = parseMediaType("application/problem+json");
+
+    private static final Logger LOG = LoggerFactory.getLogger(NakadiGateway.class);
+    private static final String SESSION_HEADER = "X-Nakadi-SessionId"; // TODO should this be X-Nakadi-StreamId?
+
+    private static final Binding<Series> ON_FAILURE_BINDING =
+            anySeries().call(dispatch(contentType(),
+                    on(PROBLEM).call(ThrowableProblem.class, propagate()),
+                    Bindings.anyContentType().call(invalidContent())));
+
+    private static ThrowingConsumer<ClientHttpResponse> invalidContent() {
+        return response -> {
+            Charset charset = Optional.ofNullable(response.getHeaders().getContentType())
+                    .map(MediaType::getCharSet)
+                    .orElse((StandardCharsets.ISO_8859_1));
+            String body = new String(ByteStreams.toByteArray(response.getBody()), charset);
+            throw new IOException("Unsupported content: " + response.getRawStatusCode() +
+                    " - " + response.getStatusText() + "\n" + response.getHeaders() + "\n" + body);
+        };
+    }
+
+    private static ThrowingConsumer<ThrowableProblem> propagate() {
+        return entity -> {
+            throw entity;
+        };
+    }
+
+    static class GatewayException extends RuntimeException {
+
+        public GatewayException(Throwable cause) {
+            super(cause);
+        }
+    }
+
+    private final Rest rest;
+
+    @Autowired
+    public NakadiGateway(final Rest rest) {
+        this.rest = rest;
+    }
+
+    public void stream(final String owningApplication, final List<String> eventTypes, final EventConsumer consumer)
```

Comment:
> Owning application should be a field/constructor argument, imho.

Yes semantically agreed. And we have to include client_group_id here.

> I'd love to have this method generic, including the EventConsumer interface.

When we substitute the generic event against specific event types, this makes sense. As long as we do not plane to translate an generic event into a specific type, I would prefer to leave it that way.